### PR TITLE
Overlapping frames

### PR DESCRIPTION
**Pull request by nipunG314**
This PR creates a simple overlapping frames sample where we first draw a triangle to a temporary Vulkan Image, and then invert its color in separate render passes.

The complete sample has been synchronized as follows:
... -> RenderPass1 -> Acquire Swapchain Image -> RenderPass2 -> Present Swapchain Image -> ...

### CODE REVIEW

File: application_sandbox/overlapping_frames/main.cpp [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+uint32_t post_frag_shader[] =
+#include "post.frag.spv"
+    ;
+
+namespace screen_model {
+#include "fullscreen_quad.obj.h"
+}
+const auto& screen_data = screen_model::model;
+
+vulkan::VkRenderPass buildRenderPass(
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:57]: Once we have a cube rendering you will need a depth buffer here as well.

File: application_sandbox/overlapping_frames/main.cpp [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+#include "vulkan_wrapper/descriptor_set_wrapper.h"
+#include "vulkan_wrapper/sub_objects.h"
+
+struct CommandTracker {
+    containers::unique_ptr<vulkan::VkCommandBuffer> command_buffer;
+    containers::unique_ptr<vulkan::VkFence> rendering_fence;
+    containers::unique_ptr<vulkan::DescriptorSet> descriptor_set;
+};
+
+uint32_t tri_vert_shader[] =
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:36]: Might be nicer to put these into a namespace:

namespace gBuffer {
    const uint32_t vert_shader[] = 
      #include ....
   const uint32_t frag_shader[] = 
     #include ....
}

Then reference as gBuffer::vert

File: application_sandbox/overlapping_frames/main.cpp [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+           VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,            // dstAccessMask
+           0                                                // dependencyFlags
+        }}
+    );
+
+    return render_pass;
+}
+
+vulkan::VulkanGraphicsPipeline buildTrianglePipeline(
+    vulkan::VulkanApplication& app,
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:104]: VulkanApplication should be passed by pointer.

File: application_sandbox/overlapping_frames/main.cpp [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+    });
+    pipeline.SetSamples(VK_SAMPLE_COUNT_1_BIT);
+    pipeline.AddAttachment();
+    pipeline.Commit();
+
+    return pipeline;
+}
+
+vulkan::VulkanGraphicsPipeline buildPostPipeline(
+    vulkan::VulkanApplication& app,
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:134]: pass by pointer please.

File: application_sandbox/overlapping_frames/main.cpp [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+    pipeline.AddAttachment();
+    pipeline.Commit();
+
+    return pipeline;
+}
+
+containers::vector<vulkan::ImagePointer> buildTempImages(
+    vulkan::VulkanApplication& app,
+    const entry::EntryData* data) {
+    containers::vector<vulkan::ImagePointer> images(data->allocator());
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:168]: images.reserve(app.swapchain_images().size());

File: application_sandbox/overlapping_frames/main.cpp [hunk 6]
```diff
@@ -0,0 +1,649 @@
...
+
+    for(int index=0; index<app.swapchain_images().size(); index++) {
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:170]: Please run clang-format on this to get the formatting right.

File: application_sandbox/overlapping_frames/main.cpp [hunk 7]
```diff
@@ -0,0 +1,649 @@
...
+    // Triangle Synchronization
+    containers::vector<vulkan::VkCommandBuffer> command_buffers_triangle(data->allocator());
+    containers::vector<vulkan::VkSemaphore> render_finished_triangle(data->allocator());
+
+    // Post Synchronization
+    containers::vector<vulkan::VkSemaphore> image_acquired(data->allocator());
+    containers::vector<vulkan::VkSemaphore> render_finished_post(data->allocator());
+    containers::vector<CommandTracker> command_trackers_post(data->allocator());
+    containers::unordered_map<uint32_t, uint32_t> progress_post(data->allocator());
+
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:437]: We should reserve the space for all values in the vector (saves re-allocation).

File: application_sandbox/overlapping_frames/main.cpp [hunk 8]
```diff
@@ -0,0 +1,649 @@
...
+        app.device()->vkAcquireNextImageKHR(
+            app.device(),
+            app.swapchain().get_raw_object(),
+            UINT64_MAX,
+            image_acquired[current_frame].get_raw_object(),
+            static_cast<VkFence>(VK_NULL_HANDLE),
+            &image_index
+        );
+
+        if (progress_post.find(image_index) != progress_post.end()) {
```
Comment (AWoloszyn) [inline application_sandbox/overlapping_frames/main.cpp:556]: Discussed offline, but this is not necessary (Also I think this is causing a hang)
There are 2 waits for any given fence, but only one signal.

Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp:587]: I removed it and the hang went away. It must have been progress_post.

File: application_sandbox/overlapping_frames/main.cpp [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+#include "support/entry/entry.h"
+#include "vulkan_core.h"
+#include "vulkan_helpers/vulkan_application.h"
+#include "vulkan_helpers/helper_functions.h"
+#include "vulkan_helpers/vulkan_model.h"
+#include "vulkan_wrapper/command_buffer_wrapper.h"
+#include "vulkan_wrapper/descriptor_set_wrapper.h"
+#include "vulkan_wrapper/sub_objects.h"
+
+// depending on the complexity of the use, maybe have proper getters and setters as opposed to raw field access.
```
Comment (nipunG314) [inline application_sandbox/overlapping_frames/main.cpp:30]: Should I implement proper getters and setters here?

File: application_sandbox/overlapping_frames/post.frag [hunk 10]
```diff
@@ -0,0 +1,28 @@
+/* Copyright 202020 Google Inc.
```
[the remark this answers is not in the document]
Comment (nipunG314) [reply application_sandbox/overlapping_frames/post.frag]: Done

File: application_sandbox/overlapping_frames/main.cpp [hunk 11]
```diff
@@ -0,0 +1,649 @@
...
+#include "post.frag.spv"
+    ;
+
+namespace screen_model {
+#include "fullscreen_quad.obj.h"
+}
+const auto& screen_data = screen_model::model;
+
+vulkan::VkRenderPass buildRenderPass(
+    vulkan::VulkanApplication& app,
```
[the remark this answers is not in the document]
Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp]: Done

File: application_sandbox/overlapping_frames/main.cpp [hunk 12]
```diff
@@ -0,0 +1,649 @@
...
+    VkImageLayout initial_layout = VK_IMAGE_LAYOUT_UNDEFINED,
```
[the remark this answers is not in the document]
Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp]: Done

File: application_sandbox/overlapping_frames/main.cpp (hunk 2 above)
Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp]: Done

File: application_sandbox/overlapping_frames/main.cpp (hunk 4 above)
Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp]: Done

File: application_sandbox/overlapping_frames/main.cpp (hunk 5 above)
Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp:220]: Done

File: application_sandbox/overlapping_frames/main.cpp (hunk 7 above)
Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp:451]: Done

File: application_sandbox/overlapping_frames/main.cpp (hunk 9 above)
Comment (AWoloszyn) [reply application_sandbox/overlapping_frames/main.cpp:66]: You are probably fine for now, If it gets any more complicated we should do SOMETHING to manage complexity.

File: application_sandbox/overlapping_frames/main.cpp (hunk 3 above)
Comment (nipunG314) [reply application_sandbox/overlapping_frames/main.cpp]: Done

**Comment by AWoloszyn** [comment]
Can you squash this into a single commit with a single sensible commit message. Then I will merge

**Comment by nipunG314** [comment]
@AWoloszyn I've squashed all the commits into one. Is this sufficient?

**Comment by AWoloszyn** [comment]
Merged

**Pull request merged**